File: epi_judge_cpp/lowest_common_ancestor.cc
#include <memory>

#include "binary_tree_node.h"
#include "test_framework/binary_tree_utils.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using std::unique_ptr;
using namespace std;

pair<int, BinaryTreeNode<int> *> myLca(const unique_ptr<BinaryTreeNode<int>> &tree,
                                       const unique_ptr<BinaryTreeNode<int>> &node0,
                                       const unique_ptr<BinaryTreeNode<int>> &node1)
{

  if (tree == nullptr)
  {
    return make_pair(0, nullptr);
  }

  // int totalNodes = 0;
  int total=0;

  if (tree==node0){
    total++;
  }

  if (tree==node1){
    total++;
  }

  auto leftRes = myLca(tree->left, node0, node1);
  auto rightRes = myLca(tree->right, node0, node1);

  if (leftRes.second != nullptr){
    return make_pair(2,leftRes.second);
  }

  if (rightRes.second != nullptr){
    return make_pair(2,rightRes.second);
  }

  int totalAmt=total+leftRes.first+rightRes.first;

  if (totalAmt==2){
    return make_pair(totalAmt,tree.get());
  }


  return make_pair(totalAmt,nullptr);



}

BinaryTreeNode<int> *Lca(const unique_ptr<BinaryTreeNode<int>> &tree,
                         const unique_ptr<BinaryTreeNode<int>> &node0,
                         const unique_ptr<BinaryTreeNode<int>> &node1)
{
  // TODO - you fill in here.

  auto res=myLca(tree,node0,node1);

  return res.second;
}
int LcaWrapper(TimedExecutor &executor,
               const unique_ptr<BinaryTreeNode<int>> &tree, int key0,
               int key1)
{
  const unique_ptr<BinaryTreeNode<int>> &node0 = MustFindNode(tree, key0);
  const unique_ptr<BinaryTreeNode<int>> &node1 = MustFindNode(tree, key1);

  auto result = executor.Run([&]
                             { return Lca(tree, node0, node1); });

  if (!result)
  {
    throw TestFailure("Result can not be nullptr");
  }
  return result->data;
}

int main(int argc, char *argv[])
{
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "tree", "key0", "key1"};
  return GenericTestMain(args, "lowest_common_ancestor.cc",
                         "lowest_common_ancestor.tsv", &LcaWrapper,
                         DefaultComparator{}, param_names);
}

File: epi_judge_python/lowest_common_ancestor.py
import functools
from typing import Optional

from binary_tree_node import BinaryTreeNode
from test_framework import generic_test
from test_framework.binary_tree_utils import must_find_node, strip_parent_link
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook

# invalid in put, can i be ancestor of myself?
# took 3 runs


def lcaBetter(tree, node0, node1):

    def lcaRec(root):
        if root is None:
            return (0, None)  # num of nodes in subtree and lca

        totalNodes = 0
        
        if root is node0:
            totalNodes += 1
            

        if root is node1:
            totalNodes += 1
            

        amtL, lcaL = lcaRec(root.left)
        amtR, lcaR = lcaRec(root.right)
        if amtL==2:
            return (amtL,lcaL)
        if amtR==2:
            return (amtR,lcaR)

        
        totalNodes+=amtL
        totalNodes+=amtR
        return (totalNodes,root)

    return lcaRec(tree)[1]
        
        

        


def lca(tree: BinaryTreeNode, node0: BinaryTreeNode,
        node1: BinaryTreeNode) -> Optional[BinaryTreeNode]:

    if tree is None:
        return None

    def lcaRec(root, pathList, target):
        if root is None:
            return False

        if root is target:
            pathList.append(target)
            return True

        pathList.append(root)
        if lcaRec(root.right, pathList, target):
            return True
        if lcaRec(root.left, pathList, target):
            return True
        pathList.pop()
        return False

    p1 = []
    p2 = []
    lcaRec(tree, p1, node0)
    lcaRec(tree, p2, node1)

    i = 0
    while i < len(p1) and i < len(p2) and p1[i] == p2[i]:
        i += 1

    return p1[i-1]


@enable_executor_hook
def lca_wrapper(executor, tree, key1, key2):
    strip_parent_link(tree)
    result = executor.run(
        functools.partial(lcaBetter, tree, must_find_node(tree, key1),
                          must_find_node(tree, key2)))

    if result is None:
        raise TestFailure('Result can\'t be None')
    return result.data


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('lowest_common_ancestor.py',
                                       'lowest_common_ancestor.tsv',
                                       lca_wrapper))
